daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package engine
+package preprocessing
+
+import java.util
+
+import com.daml.lf.data.{ImmArray, Ref}
+import com.daml.lf.language.Ast
+import com.daml.lf.speedy.SValue
+import com.daml.lf.transaction.{GenTransaction, Node, Transaction}
+import com.daml.lf.value.Value
+
+import scala.annotation.tailrec
+import scala.util.control.NoStackTrace
+
+private[engine] final class Preprocessor(compiledPackages: MutableCompiledPackages) {
+
+  import Preprocessor._
+  val transactionPreprocessor = new TransactionPreprocessor(compiledPackages)
+  import transactionPreprocessor._
+  import commandPreprocessor._
+  import valueTranslator.unsafeTranslateValue
+
+  // This pulls all the dependencies of in `typesToProcess0` and `tyConAlreadySeed0`
+  private def getDependencies(
+      typesToProcess0: List[Ast.Type],
+      tmplToProcess0: List[Ref.TypeConName],
+      tyConAlreadySeed0: Set[Ref.TypeConName] = Set.empty,
+      tmplAlreadySeed0: Set[Ref.TypeConName] = Set.empty,
```

Comment:
You spent too much time messing around with transaction seeDs lately ;)
```suggestion
      tyConAlreadySeen0: Set[Ref.TypeConName] = Set.empty,
      tmplAlreadySeen0: Set[Ref.TypeConName] = Set.empty,
```